libs/MeshKernel/src/Mesh2DToCurvilinear.cpp

```diff
@@ -0,0 +1,298 @@
+    const auto numNodes = m_mesh.GetNumNodes();
+    m_i = std::vector(numNodes, missing::intValue);
+    m_j = std::vector(numNodes, missing::intValue);
+
+    const auto firstEdge = m_mesh.m_facesEdges[initialFaceIndex][0];
+    const auto secondEdge = m_mesh.m_facesEdges[initialFaceIndex][1];
+    const auto thirdEdge = m_mesh.m_facesEdges[initialFaceIndex][2];
+    const auto fourthEdge = m_mesh.m_facesEdges[initialFaceIndex][3];
+
+    const auto firstNodeIndex = m_mesh.FindCommonNode(firstEdge, secondEdge);
+    m_i[firstNodeIndex] = 0;
+    m_j[firstNodeIndex] = 0;
+
+    const auto secondNodeIndex = m_mesh.FindCommonNode(secondEdge, thirdEdge);
+    m_i[secondNodeIndex] = 1;
+    m_j[secondNodeIndex] = 0;
+
+    const auto thirdNodeIndex = m_mesh.FindCommonNode(thirdEdge, fourthEdge);
+    m_i[thirdNodeIndex] = 1;
+    m_j[thirdNodeIndex] = 1;
+
+    const auto fourthNodeIndex = m_mesh.FindCommonNode(fourthEdge, firstEdge);
+    m_i[fourthNodeIndex] = 0;
+    m_j[fourthNodeIndex] = 1;
+
+    // 4. Grow the front using the breath first search algorithm
+    const auto numFaces = m_mesh.GetNumFaces();
+    std::vector visitedFace(numFaces, false);
+    std::queue<UInt> q;
+    q.push(initialFaceIndex);
+
+    while (!q.empty())
+    {
+        const auto face = q.front();
+        q.pop();
+
+        if (visitedFace[face])
+        {
+            continue;
+        }
+        visitedFace[face] = true;
+
+        if (m_mesh.GetNumFaceEdges(face) != geometric::numNodesInQuadrilateral)
+        {
+            continue;
+        }
+
+        const auto localNodeMapping = ComputeLocalNodeMapping(face);
+        for (UInt d = 0; d < geometric::numNodesInQuadrilateral; ++d)
+        {
+            const auto newFaceIndex = ComputeNeighbouringFaceNodes(face, localNodeMapping, d, visitedFace);
+            if (newFaceIndex != missing::uintValue)
+            {
+                q.push(newFaceIndex);
+            }
+        }
+    }
+
+    const auto matrix = ComputeCurvilinearMatrix();
+    return std::make_unique<CurvilinearGrid>(matrix, m_mesh.m_projection);
+}
+
+Eigen::Matrix<UInt, 2, 2> Mesh2DToCurvilinear::ComputeLocalNodeMapping(UInt face) const
+{
+    const auto& faceIndices = m_mesh.m_facesNodes[face];
+
+    const auto node0 = faceIndices[0];
+    const auto node1 = faceIndices[1];
+    const auto node2 = faceIndices[2];
+    const auto node3 = faceIndices[3];
+
+    const std::vector localI{m_i[node0], m_i[node1], m_i[node2], m_i[node3]};
+    const std::vector localJ{m_j[node0], m_j[node1], m_j[node2], m_j[node3]};
+    const std::vector localNodes{node0, node1, node2, node3};
+
+    const auto minI = *std::ranges::min_element(localI);
+    const auto minJ = *std::ranges::min_element(localJ);
+
+    Eigen::Matrix<UInt, 2, 2> matrix;
+    for (UInt i = 0; i < localI.size(); ++i)
+    {
+        if (localI[i] == minI && localJ[i] == minJ)
+        {
+            matrix(0, 0) = localNodes[i];
+        }
```

Comment:
In the loop below is every element of the matrix assigned a value? If not then should it be initialised with some default set of values